### ISSUE TITLE
`trial` test runs leak memory

### ISSUE THREAD

**Issue by squahtx** [matrix-org/synapse#15622]
#15334 introduced a bug where `HomeServer` instances no longer get garbage collected during `trial` test runs. As a result, memory usage grows to multiple gigabytes during test runs.

See https://github.com/matrix-org/synapse/pull/15334/files#diff-f50ec854a65082453f4086048f20c291ca6020fab973486a9da2285025394cc8R180.
When we garbage collect generation 0, `self.hs` still holds a reference to the `HomeServer`, which gets promoted to generation 1 and then never collected.

---

- [x] Fix `HomeServer`s leaking due to not being GCed.
      https://github.com/matrix-org/synapse/pull/15630
- [x] Fix `FederationRateLimiter` (and referenced `HomeServer`s) leaking due to `_rate_limiter_instances` global.
      https://github.com/matrix-org/synapse/pull/15630
- [ ] Fix config cache added in #15284 growing too large (~80 entries). `HomeServerConfig`s are pretty heavy and hold on to a lot of stuff, including Jinja templates. A suitable max size for the cache might be 8-16 entries.

**Comment by squahtx** [comment]
When I force a full GC every 50 or so gen0 GCs, a trial run no longer OOMs on my laptop. But memory usage still keeps growing and the GCs get slower and slower. So there's a separate memory leak going on.

**Comment by squahtx** [comment]
One of the leaks is due to https://github.com/matrix-org/synapse/blob/a302d3ecf75493f84fc5be616fee7d199ed12394/synapse/util/ratelimitutils.py#L89

Another "leak" is due to the config cache added in https://github.com/matrix-org/synapse/pull/15284 growing large.

**Comment by squahtx** [comment]
> Another "leak" is due to the config cache added in https://github.com/matrix-org/synapse/pull/15284 growing large.

I'm going to leave this one alone. I'd replace things with `functools.lru_cache`, except config dicts are unhashable and there's no way to give `lru_cache` a custom cache key.

**Comment by DMRobertson** [comment]
> except config dicts are unhashable

FWIW I think `immutabledict` (previously `frozendict`) is hashable?

**Comment by squahtx** [comment]
> > except config dicts are unhashable
> 
> FWIW I think `immutabledict` (previously `frozendict`) is hashable?

That's true, we could frozendictify the config dicts and use them as cache keys. I'd balk at the copying involved, but the current code does a `json.dumps` to derive the cache key every time it's called.

**Comment by squahtx** [comment]
There are two more HomeServer "leaks":
* via [`synapse.util.caches.caches_by_name`](https://github.com/matrix-org/synapse/blob/9bb2eac71962970d02842bca441f4bcdbbf93a11/synapse/util/caches/__init__.py#L39)
  `caches_by_name` -> `ResponseCache.clock` -> `Clock._reactor` -> `ThreadedMemoryReactorClock` -> ... -> `<bound method DeactivateAccountHandler._start_user_parting>` -> `DeactivateAccountHandler` -> `TestHomeServer`
* via [`synapse.util.caches.CACHE_METRIC_REGISTRY`](https://github.com/matrix-org/synapse/blob/9bb2eac71962970d02842bca441f4bcdbbf93a11/synapse/util/caches/__init__.py#L37)
  `CACHE_METRIC_REGISTRY._pre_update_hooks` -> `<bound method CacheMetric.collect>` -> `CacheMetric._cache` -> `ResponseCache.clock` -> ... (same path as above) -> `TestHomeServer`

`caches_by_name` and `CACHE_METRIC_REGISTRY` indirectly hold references to multiple homeservers from previous tests. Luckily, the number of homeservers leaked is bounded and comes out to fewer than 10 in practice.

**Comment by squahtx** [comment]
> There are ~two~ three more HomeServer "leaks":
> 
> * via [`synapse.util.caches.caches_by_name`](https://github.com/matrix-org/synapse/blob/9bb2eac71962970d02842bca441f4bcdbbf93a11/synapse/util/caches/__init__.py#L39)
>   `caches_by_name` -> `ResponseCache.clock` -> `Clock._reactor` -> `ThreadedMemoryReactorClock` -> ... -> `<bound method DeactivateAccountHandler._start_user_parting>` -> `DeactivateAccountHandler` -> `TestHomeServer`
> * via [`synapse.util.caches.CACHE_METRIC_REGISTRY`](https://github.com/matrix-org/synapse/blob/9bb2eac71962970d02842bca441f4bcdbbf93a11/synapse/util/caches/__init__.py#L37)
>   `CACHE_METRIC_REGISTRY._pre_update_hooks` -> `<bound method CacheMetric.collect>` -> `CacheMetric._cache` -> `ResponseCache.clock` -> ... (same path as above) -> `TestHomeServer`
> * via [cache resizing callbacks](https://github.com/matrix-org/synapse/blob/796a4b74823b721c72de07e45718f05e78e1565d/synapse/config/cache.py#L34).
> 
> `caches_by_name` and `CACHE_METRIC_REGISTRY` indirectly hold references to multiple homeservers from previous tests. Luckily, the number of homeservers leaked is bounded and comes out to fewer than 10 in practice.

`caches_by_name` can be turned into a [`WeakValueDictionary`](https://docs.python.org/3/library/weakref.html#weakref.WeakValueDictionary).

`CACHE_METRIC_REGISTRY._pre_update_hooks[].__self__._cache` is a lot messier. A weakref can be used for some of the caches. Other caches are simple lists or dicts, which weakref does not support.

The cache resizing callbacks are also painful to fix, even when using [`WeakMethod`](https://docs.python.org/3/library/weakref.html#weakref.WeakMethod)s.

I'm going to leave these alone.
